### PR TITLE
Allow post to post relationships using terms

### LINKED ISSUE #128: Allow relationships to be defined between posts using terms
Instead of relying on meta queries for post <--> post relationships, allow terms to be used.

### PR DESCRIPTION
**Pull request by srtfisher**
Closes #128

### CODE REVIEW

File: src/mantle/framework/database/model/relations/class-belongs-to.php [hunk 1]
```diff
@@ -52,17 +57,29 @@ public function __construct( Builder $query, Model $parent, string $foreign_key,
...
-			 * @todo Handle missing meta value better.
-			 */
-			$this->query->where( 'id', PHP_INT_MAX );
+		if ( $this->uses_terms ) {
+			$object_ids = $this->get_term_ids_for_relationship();
+
+			if ( empty( $object_ids ) ) {
+				// Prevent the query from going through.
+				// @todo Handle this better.
+				return $this->query->where( 'id', PHP_INT_MAX );
```
Comment (pattiereaves) [inline src/mantle/framework/database/model/relations/class-belongs-to.php:66]: (just curious, for my own info) is what makes it hard to handle the query like this that this function needs to return a query?

File: src/mantle/framework/database/model/relations/class-belongs-to.php [hunk 2]
```diff
@@ -96,12 +149,91 @@ public function dissociate() {
...
+
 		if ( $compare_value ) {
 			return $builder->whereMeta( $this->local_key, $compare_value, $compare ?? '' );
 		}
 
 		return $builder->whereMeta( $this->local_key, '', $compare );
 	}
+
+	/**
+	 * Retrieve a internal term for a post to post relationship.
```
Comment (pattiereaves) [inline src/mantle/framework/database/model/relations/class-belongs-to.php:168]:
```suggestion
	 * Retrieve a internal term for a post-to-post relationship.
```

File: src/mantle/framework/database/model/relations/class-belongs-to.php [hunk 3]
```diff
@@ -96,12 +149,91 @@ public function dissociate() {
...
+			}
+
+			return $insert['term_id'];
+		}
+
+		return $term->term_id;
+	}
+
+	/**
+	 * Retrieve the term slug for a post to post relationship.
```
Comment (pattiereaves) [inline src/mantle/framework/database/model/relations/class-belongs-to.php:192]:
```suggestion
	 * Retrieve the term slug for a post-to-post relationship.
```

File: src/mantle/framework/database/model/relations/class-has-one-or-many.php [hunk 4]
```diff
@@ -115,4 +145,41 @@ protected function is_post_term_relationship(): bool {
 	protected function is_term_post_relationship(): bool {
 		return $this->parent instanceof Term && $this->query instanceof Post_Query_Builder;
 	}
+
+	/**
+	 * Retrieve a internal term for a post to post relationship.
```
Comment (pattiereaves) [inline src/mantle/framework/database/model/relations/class-has-one-or-many.php:150]:
```suggestion
	 * Retrieve a internal term for a post-to-post relationship.
```

File: src/mantle/framework/database/model/relations/class-has-one-or-many.php [hunk 5]
```diff
@@ -115,4 +145,41 @@ protected function is_post_term_relationship(): bool {
...
+			}
+
+			return $insert['term_id'];
+		}
+
+		return $term->term_id;
+	}
+
+	/**
+	 * Retrieve the term slug for a post to post relationship.
```
Comment (pattiereaves) [inline src/mantle/framework/database/model/relations/class-has-one-or-many.php:177]:
```suggestion
	 * Retrieve the term slug for a post-to-post relationship.
```

File: src/mantle/framework/database/model/relations/class-relation.php [hunk 6]
```diff
@@ -17,25 +17,42 @@
 abstract class Relation {
 	use Forward_Calls;
 
+	/**
+	 * Internal taxonomy for post to post relationships.
```
Comment (pattiereaves) [inline src/mantle/framework/database/model/relations/class-relation.php:21]:
```suggestion
	 * Internal taxonomy for post-to-post relationships.
```

**Review by pattiereaves — APPROVED** [review]
🦄

File: src/mantle/framework/database/model/relations/class-belongs-to.php (hunk 1 above)
Comment (srtfisher) [reply src/mantle/framework/database/model/relations/class-belongs-to.php:66]: In this case, we need to essentially kill the query. I would have thrown an exception, but I haven't set that up yet so this is the "easiest" way to kill a query.

**Pull request merged**